src/Discord/Internal/Types/Components.hs

```diff
@@ -1,219 +1,278 @@
+import Discord.Internal.Types.Prelude (EmojiId, RoleId, toMaybeJSON)
 import Discord.Internal.Types.User (User)
-import qualified Network.HTTP.Req as R
+
+data ComponentActionRow = ComponentActionRowButton [ComponentButton] | ComponentActionRowSelectMenu ComponentSelectMenu
+  deriving (Show, Eq, Ord, Read)
+
+instance FromJSON ComponentActionRow where
+  parseJSON =
+    withObject
+      "ComponentActionRow"
+      ( \cs -> do
+          a <- cs .: "components" :: Parser Array
+          let a' = toList a
+          case a' of
+            [] -> return $ ComponentActionRowButton []
+            (c : _) ->
+              withObject
+                "ComponentActionRow item"
+                ( \v -> do
+                    t <- v .: "type" :: Parser Int
+                    case t of
+                      2 -> ComponentActionRowButton <$> mapM parseJSON a'
+                      3 -> ComponentActionRowSelectMenu <$> parseJSON c
+                      _ -> fail $ "unknown component type: " ++ show t
+                )
+                c
+      )
+
+instance ToJSON ComponentActionRow where
+  toJSON (ComponentActionRowButton bs) = object [("type", Number 1), ("components", toJSON bs)]
+  toJSON (ComponentActionRowSelectMenu bs) = object [("type", Number 1), ("components", toJSON [bs])]
 
 -- | Component type for a button, split into URL button and not URL button.
 --
 -- Don't directly send button components - they need to be within an action row.
 data ComponentButton
   = ComponentButton
-      { componentButtonCustomId :: T.Text,
+      { -- | Dev indentifier
+        componentButtonCustomId :: T.Text,
+        -- | Whether the button is disabled
         componentButtonDisabled :: Bool,
+        -- | What is the style of the button
         componentButtonStyle :: ButtonStyle,
+        -- | What is the user-facing label of the button
         componentButtonLabel :: T.Text,
+        -- | What emoji is displayed on the button
         componentButtonEmoji :: Maybe Emoji
       }
   | ComponentButtonUrl
-      { componentButtonUrl :: R.Url 'R.Https,
+      { -- | The url for the button. If this is not a valid url, everything will
+        -- break
+        componentButtonUrl :: T.Text,
+        -- | Whether the button is disabled
         componentButtonDisabled :: Bool,
+        -- | What is the user-facing label of the button
         componentButtonLabel :: T.Text,
+        -- | What emoji is displayed on the button
         componentButtonEmoji :: Maybe Emoji
       }
-  deriving (Show, Eq)
+  deriving (Show, Eq, Ord, Read)
 
-data ButtonStyle = ButtonStylePrimary | ButtonStyleSecondary | ButtonStyleSuccess | ButtonStyleDanger
-  deriving (Show, Eq, Read)
+-- | Takes the label and the custom id of the button that is to be generated.
+mkButton :: T.Text -> T.Text -> ComponentButton
+mkButton label customId = ComponentButton customId False ButtonStyleSecondary label Nothing
```

Comment:
`mkButton`, `mkSelectOption`, `mkEmoji` seem like weird names to me. `Data.Data` uses this convention though.

examples/interaction-commands uses `mkEmoji`, but just putting the string of the emoji should work as well. https://github.com/aquarial/discord-haskell/blob/master/docs/emoji.md